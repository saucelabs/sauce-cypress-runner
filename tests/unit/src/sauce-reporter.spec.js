jest.mock('fs');
jest.mock('webdriverio');
jest.mock('saucelabs');
const fs = require('fs');
const webdriverio = require('webdriverio');
const SauceLabs = require('saucelabs');
const SauceReporter = require('../../../src/sauce-reporter');

describe('SauceReporter', function () {
  const fakeRunConfig = {
    sauce: {
      metadata: {
        name: 'Fake Name',
      }
    }
  };
  describe('.prepareAssets', function () {
    it('should return a list of assets', async function () {
      fs.existsSync.mockReturnValue(true);
      fs.copyFileSync.mockReturnValue(true);
      SauceReporter.mergeVideos = jest.fn().mockImplementation(async function () {});
      const res = await SauceReporter.prepareAssets(['spec/file.test.js'], 'results/');
      expect(res).toMatchSnapshot();
    });
    it('should return an empty list of assets if files not found (addresses DEVX-273)', async function () {
      fs.existsSync.mockReturnValue(false);
      fs.copyFileSync.mockReturnValue(true);
      fs.mkdtempSync.mockReturnValue('tmp/folder');
      const res = await SauceReporter.prepareAssets('spec/file', 'results/');
      expect(res).toEqual([]);
    });
  });
  describe('.sauceReporter', function () {
    let prepareAssetsSpy, uploadJobAssetsSpy;
    beforeEach(function () {
      webdriverio.remote.mockImplementation(function () {});
      prepareAssetsSpy = jest.spyOn(SauceReporter, 'prepareAssets');
      // eslint-disable-next-line require-await
      uploadJobAssetsSpy = jest.fn().mockImplementation(async () => ({errors: ['some fake error']}));
      SauceLabs.default.mockImplementation(function () {
        // eslint-disable-next-line require-await
        this.listJobs = async () => ({
          jobs: [{id: 'a'}, {id: 'b'}]
        });
        this.uploadJobAssets = uploadJobAssetsSpy;
        this.updateJob = async () => {};
      });
    });
    it('should call uploadJobAssets on SauceLabs api', async function () {
      prepareAssetsSpy.mockReturnValue(['asset/one', 'asset/two']);
      await SauceReporter.sauceReporter(fakeRunConfig, 'build', 'browser', ['asset/one', 'asset/two'], 0);
      expect(uploadJobAssetsSpy.mock.calls).toEqual([
        ['a', {'files': ['asset/one', 'asset/two']}]
      ]);
    });
    it('should output err when upload failed', async function () {
      let consoleErrorSpy = jest.spyOn(global.console, 'error');
      prepareAssetsSpy.mockReturnValue(['asset/one', 'asset/two']);
      expect(await SauceReporter.sauceReporter(fakeRunConfig, 'build', 'browser', ['asset/one', 'asset/two'], 0)).toBeUndefined();
      expect(uploadJobAssetsSpy.mock.calls).toEqual([
        ['a', {'files': ['asset/one', 'asset/two']}]
      ]);
      expect(consoleErrorSpy.mock.calls).toEqual([['some fake error']]);
    });
    it('should not push assets when no sessionId from SauceLabs API', async function () {
      SauceLabs.default.mockImplementation(function () {
        // eslint-disable-next-line require-await
        this.listJobs = async () => ({
          jobs: []
        });
        this.uploadJobAssets = uploadJobAssetsSpy;
        this.updateJob = async () => {};
      });

      prepareAssetsSpy.mockReturnValue(['asset/one', 'asset/two']);
      expect(await SauceReporter.sauceReporter(fakeRunConfig, 'build', 'browser', ['asset/one', 'asset/two'], 0)).toBeDefined();
    });
  });
});
