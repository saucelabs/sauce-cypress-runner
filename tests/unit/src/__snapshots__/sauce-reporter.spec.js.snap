// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SauceReporter .prepareAssets should return a list of assets 1`] = `
Array [
  "console.log",
  "results/spec/file.test.js.mp4",
  "results/spec/file.test.js.json",
  "results/spec/file.test.js.xml",
  "results/video.mp4",
]
`;

exports[`SauceReporter .sauceReporter should create job via global data store 1`] = `
Array [
  Array [
    Object {
      "acl": Array [
        Object {
          "type": "username",
          "value": "fake-user",
        },
      ],
      "attributes": Object {
        "browser": "browser",
        "browser_version": "*",
        "commands_not_successful": 1,
        "container": false,
        "devx": true,
        "log_url": "test",
        "os": "test",
        "performance_enabled": "true",
        "public": "team",
        "record_logs": true,
        "record_mp4": "true",
        "record_screenshots": "true",
        "record_video": "true",
        "video_url": "test",
      },
      "live": false,
      "metadata": Object {},
      "name": "Fake Name - build",
      "platform": "webdriver",
      "source": "vdc",
      "status": "complete",
      "tags": undefined,
    },
  ],
]
`;

exports[`SauceReporter .sauceReporter should output err when upload failed 1`] = `
Array [
  Array [
    "some fake error",
  ],
]
`;
